plugins/processors/snmp_lookup/lookup.go

```diff
@@ -0,0 +1,326 @@
+	SecLevelTag     string `toml:"sec_level_tag"`
+	AuthProtocolTag string `toml:"auth_protocol_tag"`
+	AuthPasswordTag string `toml:"auth_password_tag"`
+	PrivProtocolTag string `toml:"priv_protocol_tag"`
+	PrivPasswordTag string `toml:"priv_password_tag"`
+	ContextNameTag  string `toml:"context_name_tag"`
+
+	CacheSize       int             `toml:"max_cache_entries"`
+	ParallelLookups int             `toml:"max_parallel_lookups"`
+	Ordered         bool            `toml:"ordered"`
+	CacheTTL        config.Duration `toml:"cache_ttl"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	cache    *expirable.LRU[string, tagMap]
+	parallel parallel.Parallel
+	sigs     signalMap
+	lock     sync.Mutex
+	table    si.Table
+
+	getConnection getConnectionFunc
+
+	translator si.Translator
+}
+
+const (
+	defaultCacheSize       = 100
+	defaultCacheTTL        = config.Duration(8 * time.Hour)
+	defaultParallelLookups = 100
+	minRetry               = 5 * time.Minute
+	orderedQueueSize       = 10_000
+)
+
+func (*Lookup) SampleConfig() string {
+	return sampleConfig
+}
+
+func (l *Lookup) SetTranslator(name string) {
+	if name != "gosmi" {
+		l.Log.Warnf("unsupported agent.snmp_translator value %q, some features might not work", name)
+	}
+
+	l.Translator = name
+}
+
+func (l *Lookup) Init() (err error) {
+	l.sigs = make(signalMap)
+	l.getConnection = l.getConnectionNoMock
+
+	if _, err = snmp.NewWrapper(l.ClientConfig); err != nil {
+		return fmt.Errorf("parsing SNMP client config: %w", err)
+	}
+
+	switch l.Translator {
+	case "", "gosmi":
+		if l.translator, err = si.NewGosmiTranslator(l.Path, l.Log); err != nil {
+			return fmt.Errorf("loading translator: %w", err)
+		}
+	case "netsnmp":
+		l.translator = si.NewNetsnmpTranslator()
+	default:
+		return fmt.Errorf("invalid agent.snmp_translator value %q", l.Translator)
+	}
+
+	return l.initTable()
+}
+
+func (l *Lookup) initTable() error {
+	l.table.Name = "lookup"
+	l.table.IndexAsTag = true
+	l.table.Fields = make([]si.Field, len(l.Tags))
+	for i, tag := range l.Tags {
+		tag.IsTag = true
+		l.table.Fields[i] = tag
+	}
+
+	return l.table.Init(l.translator)
+}
+
+func (l *Lookup) Start(acc telegraf.Accumulator) error {
+	l.cache = expirable.NewLRU[string, tagMap](l.CacheSize, nil, time.Duration(l.CacheTTL))
+	if l.Ordered {
+		l.parallel = parallel.NewOrdered(acc, l.addAsync, orderedQueueSize, l.ParallelLookups)
+	} else {
+		l.parallel = parallel.NewUnordered(acc, l.addAsync, l.ParallelLookups)
+	}
+	return nil
+}
+
+func (l *Lookup) Add(metric telegraf.Metric, _ telegraf.Accumulator) error {
+	l.parallel.Enqueue(metric)
+	return nil
+}
+
+func (l *Lookup) addAsync(metric telegraf.Metric) []telegraf.Metric {
+	agent, ok := metric.GetTag(l.AgentTag)
+	if !ok {
+		l.Log.Warn("Agent tag missing")
+		return []telegraf.Metric{metric}
+	}
+
+	index, ok := metric.GetTag(l.IndexTag)
+	if !ok {
+		l.Log.Warn("Index tag missing")
+		return []telegraf.Metric{metric}
+	}
+
+	// Check cache
+	l.lock.Lock()
```

Comment:
It is to prevent a second thread thinking there's a cache miss and no tread is already doing the gathering. To prevent 2 threads building the lookup table from the same agent. So "race condition".

Reply:
@srebhan can this be resolved or is it still not clear enough?

Reply:
Resolved with [9eff6fc](https://github.com/influxdata/telegraf/pull/14223/commits/9eff6fc671057ff4bbf2946ab40eb230b0d61894)